src/Http/Middleware/SessionCsrfProtectionMiddleware.php

```diff
@@ -148,6 +148,55 @@ public function skipCheckCallback(callable $callback)
         return $this;
     }
 
+    /**
+     * Apply entropy to a CSRF token
+     *
+     * To avoid BREACH apply a random salt value to a token
+     * When the token is compared to the session the token needs
+     * to be unsalted.
+     *
+     * @param string $token The token to salt.
+     * @return string The salted token with the salt appended.
+     */
+    public function saltToken(string $token): string
+    {
+        $length = strlen($token);
+        $salt = Security::randomString($length);
+        $salted = '';
+        for ($i = 0; $i < $length; $i++) {
+            // XOR the token and salt together so that we can reverse it later.
+            $salted .= chr(ord($token[$i]) ^ ord($salt[$i]));
```

Comment:
Don't want to encrypt it with the salt?

Reply:
Result of XOR will be likely codepoint of binary character, so `chr()` will make binary string. Binary strings are unwanted in HTTP headers or forms.

`$token` and `$salt` are hex strings, so they should be hex unencoded first, then hex re-encoded after XOR operation.

So `unsaltToken()` should also check if token is valid hexadecimal string before hex decoding.

Reply:
Good point about the binary data being bad in HTTP contexts. I'll make the suggested changes.

Reply:
I switched to using base64 encoding as it was simpler to use and more compact than hex encoding is.